noir-projects/aztec-nr/aztec/src/encrypted_logs/log_assembly_strategies/default_aes128/note.nr

```diff
@@ -0,0 +1,557 @@
+    // *****************************************************************************
+
+    // Already done for this strategy: AES expects bytes.
+
+    // *****************************************************************************
+    // Encrypt the plaintext
+    // *****************************************************************************
+
+    let (sym_key, iv) = derive_aes_symmetric_key_and_iv_from_ecdh_shared_secret_using_sha256(
+        ciphertext_shared_secret,
+    );
+
+    let ciphertext_bytes = aes128_encrypt(final_plaintext_bytes, iv, sym_key);
+
+    // |full_pt| = |pt_length| + |pt|
+    // |pt_aes_padding| = 16 - (|full_pt| % 16)
+    // or... since a % b is the same as a - b * (a // b) (integer division), so:
+    // |pt_aes_padding| = 16 - (|full_pt| - 16 * (|full_pt| // 16))
+    // |ct| = |full_pt| + |pt_aes_padding|
+    //      = |full_pt| + 16 - (|full_pt| - 16 * (|full_pt| // 16))
+    //      = 16 + 16 * (|full_pt| // 16)
+    //      = 16 * (1 + |full_pt| // 16)
+    assert(ciphertext_bytes.len() == 16 * (1 + ((N * 32) + 64) / 16));
+
+    // *****************************************************************************
+    // Compute the header ciphertext
+    // *****************************************************************************
+
+    let contract_address = context.this_address();
+    let contract_address_bytes = contract_address.to_field().to_be_bytes::<32>();
+
+    let mut header_plaintext: [u8; 32 + 2] = [0; 32 + 2];
+    for i in 0..32 {
+        header_plaintext[i] = contract_address_bytes[i];
+    }
+    let offset = 32;
+    let ciphertext_bytes_length = ciphertext_bytes.len();
+    header_plaintext[offset] = (ciphertext_bytes_length >> 8) as u8;
+    header_plaintext[offset + 1] = ciphertext_bytes_length as u8;
+
+    // TODO: this is insecure and wasteful:
+    // "Insecure", because the esk shouldn't be used twice (once for the header,
+    // and again for the proper ciphertext) (at least, I never got the
+    // "go ahead" that this would be safe, unfortunately).
+    // "Wasteful", because the exact same computation is happening further down.
+    // I'm leaving that 2nd computation where it is, because this 1st computation
+    // will be imminently deleted, when the header logic is deleted.
+    let (sym_key, iv) = derive_aes_symmetric_key_and_iv_from_ecdh_shared_secret_using_sha256(
+        ciphertext_shared_secret,
+    );
+
+    // Note: the aes128_encrypt builtin fn automatically appends bytes to the
+    // input, according to pkcs#7; hence why the output `header_ciphertext_bytes` is 16
+    // bytes larger than the input in this case.
+    let header_ciphertext_bytes = aes128_encrypt(header_plaintext, iv, sym_key);
+    // I recall that converting a slice to an array incurs constraints, so I'll check the length this way instead:
+    assert(header_ciphertext_bytes.len() == HEADER_CIPHERTEXT_SIZE_IN_BYTES);
```

Comment:
I don't get this, dones't `aes128_encrypt` return an array?

Reply:
Ohhh, it didn't return an array when I began the PR before xmas, but then the Noir team changed it to return an array.